app/services/structured_fact_importer.rb

```diff
@@ -19,6 +19,15 @@ def normalize_recurse(memo, facts, prefix = '')
     memo
   end
 
+  def preload_fact_names
+    super.tap do |fact_names|
+      # Fetch compose values, generating {NAME => [ID, COMPOSE]}, avoiding loading the entire model
+      fact_name_class.where(:id => fact_names.values).reorder('').pluck(:name, :compose).each do |attrs|
+        fact_names[attrs[0]] = [fact_names[attrs[0]], attrs[1]]
+      end
+    end
+  end
```

Comment:
This is ugly and unreadable, won't perform well. Can't you simply do:

    Hash[fact_name_class.where(:id => fact_names.values).reorder('').pluck(:name, :id, :compose).map{|x| [x.first, [x.second, x.third]]}]